crm_rma_prodlot_invoice/models/sale_order.py

```diff
@@ -0,0 +1,91 @@
+
+from openerp import api, models
+
+
+class SaleOrder(models.Model):
+
+    _inherit = 'sale.order'
+
+    @api.model
+    def action_done(self):
+        res = super(SaleOrder, self).action_done()
+        prodlot_obj = self.env['stock.production.lot']
+        for order in self:
+            move_ids = [move for pick in order.picking_ids
+                        for move in pick.move_lines]
+            invoice_lines = [line for inv in order.invoice_ids
+                             for line in inv.invoice_line]
+            if invoice_lines and move_ids:
+                for inv_line in invoice_lines:
+                    for mov in move_ids:
+                        quants = mov.quant_ids
+                        if quants:
+                            lots = list(set([qua.lot_id for qua in quants]))
+                            for lot in lots:
+
+                                lots_with_inv_line = prodlot_obj.\
+                                    search([('invoice_line_id', '=',
+                                            inv_line.id)])
+
+                                if not lot.invoice_line_id and \
+                                        inv_line.product_id.id == \
+                                        lot.product_id.id and \
+                                        len(lots_with_inv_line) < \
+                                        inv_line.quantity:
+                                    lot.write({'invoice_line_id': inv_line.id})
+        return res
+
+    @api.multi
+    def action_ship_create(self):
+        """
+        Set stock moves to invoice lines that matches with
+        product and product quantity
+        """
+        res = super(SaleOrder, self).action_ship_create()
+        prodlot_obj = self.env['stock.production.lot']
+        move_ids = [move for pick in self.picking_ids
+                    for move in pick.move_lines]
+        invoice_lines = [line for inv in self.invoice_ids
+                         for line in inv.invoice_line]
+
+        if invoice_lines and move_ids:
+            for inv_line in invoice_lines:
+                for mov in move_ids:
+                    quants = mov.quant_ids
+                    if quants:
+                        lot_ids = [qua.lot_id.id for qua in quants]
+                        lot_ids = list(set(lot_ids))
+                        lots = prodlot_obj.browse(lot_ids)
+                        for lot in lots:
+
+                            lots_with_inv_line = prodlot_obj.\
+                                search([('invoice_line_id', '=',
+                                        inv_line.id)])
+
+                            if not lot.invoice_line_id and \
+                                    inv_line.product_id.id == \
+                                    lot.product_id.id and \
+                                    len(lots_with_inv_line) < \
+                                    inv_line.quantity:
```

Comment:
The lots with the same invoice line must be less to the quantity of product in that invoice line, because,
The amount of products related with the same invoice line can not be overcome by the quantity in the line. The stock.production.lot model has a invoice_line_id field.